### PR TITLE
[WiP] SG-268: Add DBUS API to access control

### PR DESCRIPTION
**Pull request by przemyslawfierek-silvair**

### CODE REVIEW

File: mesh/mesh-adapter-tcpserver.c [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+	if (!jcfg)
+		return false;
+
+	if (!load_acl(jcfg, acl))
+		return false;
+
+	return true;
+}
+
+static bool add_token(json_object *jobj, uint64_t token)
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-adapter-tcpserver.c]: I think you can keep the token simply as a uint64 number, without hex encoding.

File: mesh/mesh-io-tcpserver.c [hunk 2]
```diff
@@ -42,17 +42,14 @@
 
 #include "src/shared/io.h"
 
-#include "mesh/crypto.h"
 #include "mesh/mesh-io.h"
 #include "mesh/mesh-io-api.h"
 #include "mesh/mesh-io-tcpserver.h"
 #include "mesh/silvair-io.h"
+#include "mesh/mesh-adapter-manager.h"
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: I'd rename this to just mesh-adapter.

File: mesh/mesh-adapter-tcpserver.c (hunk 1 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-adapter-tcpserver.c]: `uint64_t` is not supported by JSON-c library. It is possible to store token as `int64_t` or hexstr. Should I change it to store it as `int64_t`?

File: mesh/mesh-adapter-tcpserver.c [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+	size_t len;
+
+	if (!json_object_object_get_ex(jobj, key, &jval))
+		return 0;
+
+	str = json_object_get_string(jval);
+	if (!str)
+		return 0;
+
+	data = l_util_from_hexstring(str, &len);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-adapter-tcpserver.c]: There's a poorly named function in util.h: `str2hex`. See how keys are handled in mesh-config-json, might be of interest.

File: mesh/mesh-adapter-tcpserver.c (hunk 1 above)
Comment (michallowasrzechonek-silvair) [reply mesh/mesh-adapter-tcpserver.c]: Alright, let's leave it as hexstring then.

File: mesh/mesh-adapter-api.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ */
+#include <ell/dbus.h>
+#include <json-c/json.h>
+
+struct mesh_adpt;
+struct mesh_adpt_private;
+
+
+typedef bool (*mesh_adpt_api_reg_dbus_ifaces)(struct l_dbus *bus);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-adapter-api.h]: These names suck ;)

Comment (michallowasrzechonek-silvair) [reply mesh/mesh-adapter-api.h]: Also, we should be taking adapter as the first argument, following 'object oriented C' conventions.

File: mesh/mesh-io.c [hunk 5]
```diff
@@ -194,3 +195,13 @@ bool mesh_io_send_cancel(struct mesh_io *io, const uint8_t *pattern,
 
 	return false;
 }
+
+bool mesh_io_dbus_init(struct l_dbus *dbus)
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io.c]: This file implements `mesh_io` object, so all public API functions should take a pointer to "this" object.

You can make the API cleaner by changing the signature to `bool mesh_io_dbus_init(struct mesh_io *io, struct l_dbus *dbus)` and calling this function from `mesh_dbus_init` (passing `mesh.io` as the first argument) instead of calling it from `dbus_init`.

File: mesh/silvair-io.c [hunk 6]
```diff
@@ -752,3 +754,10 @@ void silvair_io_destroy(struct silvair_io *io)
 	if (io->tls_conn)
 		SSL_free(io->tls_conn);
 }
+
+void silvair_io_close(struct silvair_io *io)
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c:761]: Is this related to connection API? I think it *at least* belongs in a separate commit, if not another PR.

File: mesh/tcpserver-acl.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+struct tcpserver_acl;
+
+
+typedef bool (*on_acl_entry_changed)(struct tcpserver_acl *acl,
+					uint8_t *identity, void *user_data);
+
+
+struct tcpserver_acl *tcpserver_acl_new(const char *config_path,
+			on_acl_entry_changed on_acl_entry_created_callback,
```
Comment (michallowasrzechonek-silvair) [inline mesh/tcpserver-acl.h]: Meh, just add another argument to the callback to indicate operation.

Also, I don't get the difference between 'created' and 'loaded' - it shouldn't matter if the entry comes from a file or from a dbus call, no? So maybe just a `bool` argument to indicate creation/removal would be enough?

File: mesh/tcpserver-acl.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+			on_acl_entry_changed on_acl_entry_loaded_callback,
+			on_acl_entry_changed on_acl_entry_removed_callback,
+			void *user_data);
+
+void tcpserver_acl_destroy(struct tcpserver_acl *acl);
+
+bool tcpserver_acl_register_dbus_interface(
```
Comment (michallowasrzechonek-silvair) [inline mesh/tcpserver-acl.h]: I don't think this needs to be a separate public function. Just rename `tcpserver_acl_create_dbus_object` to `tcpserver_acl_dbus_init` and move interface registration there.

Note that you call these 2 one after another anyway, in `tcpserver_io_dbus_init`, and we don't support multiple instances. This can be changed when/if we get to that point.

File: mesh/tcpserver-acl.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+				struct tcpserver_acl *acl, struct l_dbus *bus);
+
+bool tcpserver_acl_create_dbus_object(struct tcpserver_acl *acl,
+					struct l_dbus *bus, const char *path);
+
+size_t tcpserver_acl_copy_psk(struct tcpserver_acl *acl,
```
Comment (michallowasrzechonek-silvair) [inline mesh/tcpserver-acl.h]: I would call it 'tcpserver_acl_psk_get`.

File: mesh/tcpserver-acl.c [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+		goto fail;
+
+	if (!read_hexstr_data(jobj, "identity", entry->identity, IDENTITY_LEN))
+		goto fail;
+
+	if (!read_hexstr_data(jobj, "psk", entry->psk, PSK_LEN))
+		goto fail;
+
+	return entry;
+	fail:
```
Comment (michallowasrzechonek-silvair) [inline mesh/tcpserver-acl.c]: indent

File: mesh/mesh-io-tcpserver.c [hunk 11]
```diff
@@ -61,24 +63,29 @@ struct mesh_io_private {
...
-	uint8_t				*uuid;
-	uint8_t				*dev_key;
-	uint8_t				*net_key;
+	const char		*dbus_path;
+	struct l_dbus		*dbus;
+	struct tcpserver_acl	*acl;
 };
 
+struct tcpserver_conn {
+	uint8_t			identity_buf[IDENTITY_BUF_LEN];
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: I'd call it `identity_str`

File: mesh/mesh-io-tcpserver.c [hunk 12]
```diff
@@ -203,19 +233,28 @@ static void io_error_callback(struct silvair_io *silvair_io)
...
 		return;
 
-	if (!l_queue_remove(mesh_io->pvt->client_io, silvair_io)) {
-		perror("l_queue_remove() error");
-		abort();
-	}
+	conn = l_queue_find(mesh_io->pvt->conns,
+				match_tcpserver_conn_by_silvair_io, silvair_io);
+	if (conn)
+		conn->io = NULL;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c:264]: Shouldn't we `conn->conn_stat->connected = false` here?

Also, this is the point where you should emit `PropertiesChanged` signal.

File: mesh/mesh-io-tcpserver.c [hunk 13]
```diff
@@ -314,51 +351,42 @@ static void tcpserver_io_init_done(void *user_data)
...
-
-	// psk = k1(dev_key, id, "idpsk")
-	if (!mesh_crypto_k1(pvt->dev_key, id, k1_info, sizeof(k1_info), psk))
-		goto done;
+	conn->io = silvair_io;
 
-	// psk is constant length
-	psk_len = 16;
+	if (conn->conn_stat)
+		conn->conn_stat->connected = true;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: Same: emit `PropertiesChanged` here.

File: mesh/mesh-io-tcpserver.c [hunk 14]
```diff
@@ -314,51 +351,42 @@ static void tcpserver_io_init_done(void *user_data)
...
-	id_str = l_util_hexstring_upper(id, sizeof(id));
-
-	// compare str format 'identity' and 'id_str'
-	if (!strcmp(identity, id_str))
-		goto done;
-
-	// psk = k1(dev_key, id, "idpsk")
-	if (!mesh_crypto_k1(pvt->dev_key, id, k1_info, sizeof(k1_info), psk))
-		goto done;
+	conn->io = silvair_io;
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c:399]: This part is nice. Exactly what I was hoping for :heart:.

File: mesh/mesh-io-tcpserver.c [hunk 15]
```diff
@@ -388,35 +416,53 @@ static bool tls_ctx_init(struct mesh_io_private *pvt)
...
+	struct tcpserver_conn	*conn = tcpserver_conn_new(identity_buf);
 
-	pvt->uuid = l_util_from_hexstring(argv[1], &len);
+	if (!l_queue_push_tail(mesh_io->pvt->conns, conn))
+		return false;
 
-	if (!pvt->uuid || len != UUID_LEN) {
-		l_error("Invalid UUID length");
+	if (!hex2str(identity_buf, IDENTITY_BUF_LEN, identity,
+							     sizeof(identity)))
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: I'm somewhat confused if `tcpserver_conn` holds binary or textual representation of peer's identity.

I think we should keep the binary representation in memory and convert strings incoming from TLS handshake only to find a corresponding PSK.

File: mesh/mesh-io-tcpserver.c [hunk 16]
```diff
@@ -388,35 +416,53 @@ static bool tls_ctx_init(struct mesh_io_private *pvt)
...
+			match_tcpserver_conn_identity_buf, identity_buf);
+	if (!conn)
 		return false;
-	}
 
-	return true;
+	tcpserver_conn_destroy(conn);
+
+	// TODO(prfi):
+	l_queue_remove(mesh_io->pvt->conns, conn);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: Why is this a `TODO`? Looks fine?

**Review by michallowasrzechonek-silvair — CHANGES_REQUESTED** [review]
Way better!

File: mesh/mesh-io-tcpserver.c (hunk 16 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-io-tcpserver.c]: There is `return false;` to avoid overwriting configuration file (just for test purposes).

File: mesh/silvair-io.c (hunk 6 above)
Comment (przemyslawfierek-silvair) [reply mesh/silvair-io.c:761]: How about create separate PR with this functionality to master and then rebase?

File: mesh/mesh-io-tcpserver.c (hunk 15 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-io-tcpserver.c]: `tcpserver_conn` keeps binary representation of peer's identity. I've found that in rest of code, binary representation is called with `_buf` postfix, so I've decided to call it similar.

File: mesh/mesh-io-tcpserver.c (hunk 11 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-io-tcpserver.c]: `identity` here is kept in the binary representation.

File: mesh/mesh-io-tcpserver.c (hunk 12 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-io-tcpserver.c]: `conn_stat` will provide functions that will emit signals (as we discussed). I've put it here just for test purpose.

File: mesh/mesh-io-tcpserver.c (hunk 13 above)
Comment (przemyslawfierek-silvair) [reply mesh/mesh-io-tcpserver.c]: conn_stat will provide functions that will emit signals (as we discussed). I've put it here just for test purpose.

File: mesh/tcpserver-acl.h (hunk 7 above)
Comment (przemyslawfierek-silvair) [reply mesh/tcpserver-acl.h]: Let's discuss this.

File: mesh/conn-stat.c [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+	l_free(conn_stat);
+}
+
+void conn_stat_set_connected(struct conn_stat *conn_stat, bool connected)
+{
+	if (conn_stat->connected == connected)
+		return;
+
+	if (connected) {
+		conn_stat->connected = false;
```
Comment (michallowasrzechonek-silvair) [inline mesh/conn-stat.c]: Seems redundant?

File: mesh/conn-stat.c [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+		conn_stat->tx_msgs_cnt = 0;
+		conn_stat->rx_msgs_cnt = 0;
+		conn_stat->last_tx_msg_timestamp = 0;
+		conn_stat->last_rx_msg_timestamp = 0;
+	}
+
+	conn_stat->connected = connected;
+
+	l_dbus_property_changed(conn_stat->dbus,
```
Comment (michallowasrzechonek-silvair) [inline mesh/conn-stat.c]: formatting ;P

File: mesh/mesh-io-tcpserver.c [hunk 19]
```diff
@@ -187,6 +195,12 @@ static void process_rx(struct silvair_io *silvair_io, int8_t rssi,
 
 		silvair_io_keep_alive_wdt_refresh(silvair_io);
 
+	conn = l_queue_find(mesh_io->pvt->conns,
+			    match_tcpserver_conn_by_silvair_io, silvair_io);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c:200]: formatting

**Comment by michallowasrzechonek-silvair** [comment]
```
WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#31: FILE: mesh/mesh-io-api.h:48:
+^Imesh_io_dbus_init_t ^Idbus_init;$

total: 0 errors, 1 warnings, 52 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Commit 9709d88155d7 ("mesh: Add D-Bus initialization API to 'mesh-io'") has style problems, please review.
```

**Comment by michallowasrzechonek-silvair** [comment]
Double check the style (see above, do checkpatch on *every* commit!), think about shortening a few function names, and we're ready to merge. Nice work.

**Comment by michallowasrzechonek-silvair** [comment]
I think you broke `--io=hci0`:

```
$ sudo ./mesh/bluetooth-meshd --io=hci0
2020-03-10 11:36:10.677       main.c:125  ready_callback             D-Bus ready
2020-03-10 11:36:10.677       main.c:106  request_name_callback      Request name success
2020-03-10 11:36:10.677       mesh.c:263  mesh_init                  Loading node configuration from /var/lib/bluetooth/mesh
...
2020-03-10 11:36:10.792       mesh.c:869  mesh_dbus_init             Added Network Interface on /org/bluez/mesh
2020-03-10 11:36:10.793       main.c:98   mesh_ready_callback        Failed to initialize mesh D-Bus resources
```
